compaction/task_manager_module.cc

```diff
@@ -266,6 +266,7 @@ future<tasks::task_manager::task::progress> compaction_task_impl::get_progress(c
 }
 
 future<> major_keyspace_compaction_task_impl::run() {
+    co_await replica::database::flush_on_all_shards(_db);
```

Comment:
What prevents a mutation to be added to commitlog right after this call?

Is the problem one of correctness (a tombstone will be incorrectly purged) or efficiency (compaction will detect the problem and avoid purging the tombstone)?

Flushing all tables can be a bad tradeoff in a cluster with many tables.

Reply:
Nothing is failing. But since we now take potential data in commit log into account when deciding what time boundaries exist for GC, tests that assume they can add data, wait X time, compact, and voila, GC has happened won't always work, since unless commitlog is properly emptied as well, compaction might not (yet) be able to apply tombstones etc. 

Tests can handle this simply by calling nodetool flush instead, this will do the same as is done here in this patch (i.e. clear out active segment so memtable flush can empty all segments in a test). I modified a bunch of tests, and sent a patch that did the same with the test mentioned in the issue. 

This PR does the same thing unconditionally before every (major) compaction instead.

Reply:
I see, so it was really a test problem.

Now, we need to consider if a flush of all tables is worth the additional tombstones collected, yes?

Reply:
Yes. I am on the fence. But I agree with the author that it might be worth it to maintain behaviour for tests. Major compaction should not be that frequent, should it?
The other option is to patch tests that make assumptions. 

If the major compaction task could decide which CF:s it intends to work on before starting, we could also limit the flushes and CL segment discards of course. Perhaps a later improvement.

Reply:
> I see, so it was really a test problem.

A test is an application.

> Nothing is failing. But since we now take potential data in commit log into account when deciding what time boundaries exist for GC, tests that assume they can add data, wait X time, compact, and voila, GC has happened won't always work, since unless commitlog is properly emptied as well, compaction might not (yet) be able to apply tombstones etc.

Compaction purges tombstones if they meet 2 conditions:
1. They are eligible for purging based on the tombstone_gc mode (either gc_grace_seconds passed since the tombstone was written, or a successful repair included them in tombstone_gc_mode=repair)
2. The tombstones shadow no data in other sstables, based on bloom filter matching, and timestamp ranges of the matched sstables.

The reason tables are flushed in 64ec1c6ec64291064678e3818e3ffdee4294dbfe (#9514) is the 2nd point above - since mutations in the memtable are not present in sstables we flush the memtables to use the sstables shadowing logic.

This change adds to that "flushing" of the commitlog so to ensure the commitlog won't resurrect data (or tombstones) after tombstone_gc.

To avoid it, we'll to maintain some metadata in core like a bloom filter for the commitlog (can be done per segment) and min/max timestamp again for each segment, so that compaction will postpone tombstone purging until it can verify that the tombstones do not potentially shadow any data in the commitlog.

We can do that as a follow-up as an optimization to this patch.
Ans a similar approach can be taken for the memtable to save their flushing.

A big advantage of the latter approach is that it will work for tombstone_gc also in regular compaction.

Reply:
No, this change does nothing to handle CL resurrection. That was handled in #15060, which ensures we don't try to GC things if there are potential CL entries that would collide with said GC. What this patch does it try to eliminate remaining CL data so that a compaction can be fully realized earlier. Equal results would been had from, for example, writing a bunch of unrelated data and thus similarly switching active CL segment. 

The "problem" with #15060, for tests and others that assume that only time waited affects GC-ability, is that whatever writes/deletes etc done by the test/program up until the waiting will end up with a portion of this data in a CL segment that is still active (allocating), and thus not eligible for deletion, even though memtable flush might clear out data for it. 
As a workaround explicit nodetool flush (the magic wand of all tests) nowadays forces rotating active CL segment as well. 

The above change here basically adds a nodetool flush to all major compactions.

Reply:
I see, so the corrected system behavior broke the test.  That makes sense.

> Now, we need to consider if a flush of all tables is worth the additional tombstones collected, yes?

Yes, I think it's worth it.
Major compaction is a rare event, and it is used many times by FEs and customers as a way to purge tombstone when too many expired tombstone were collected and regular compaction is not able to purge them for some reason, so this patch will make that step more complete.
We've seen cases in the field were lightly written tables pinned commitlog segments for days if not weeks and that might prevent tombstone gc to do its thing.
Requiring the user or test developer to flush all tables prior to major compaction is not-trivial and prone to human error, hence the automation in this patch should be preferred.

Reply:
Ok. I agree.

But we need some test here.

Reply:
The `cleanup_tet.py::TestCleanup::test_cluster_restore_no_resurrection` dtest consistently fails without this patch and consistently passes with it.